### PR TITLE
Extensible PythonPlugin:SearchPaths

### PR DESCRIPTION
**Pull request by mitchute**
### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: idd/Energy+.idd.in [hunk 1]
```diff
@@ -108583,6 +108583,7 @@ PythonPlugin:SearchPaths,
        \memo that point to libraries of plugin scripts.
        \unique-object
        \min-fields 1
+       \extensible:1
```
Comment (Myoldmopar) [inline idd/Energy+.idd.in:108586]: Marked as extensible :heavy_check_mark:

File: idd/Energy+.idd.in [hunk 2]
```diff
@@ -108605,6 +108606,7 @@ PythonPlugin:SearchPaths,
        \default Yes
    A4, \field Search Path 1
        \type alpha
+       \begin-extensible
```
Comment (Myoldmopar) [inline idd/Energy+.idd.in:108609]: Begins extensible on the first search path :heavy_check_mark:

File: idd/Energy+.idd.in [hunk 3]
```diff
@@ -108615,7 +108617,22 @@ PythonPlugin:SearchPaths,
...
+   A10,\field Search Path 7
+       \type alpha
+       \retaincase
+   A11,\field Search Path 8
+       \type alpha
+       \retaincase
+   A12,\field Search Path 9
+       \type alpha
+       \retaincase
+   A13;\field Search Path 10
```
Comment (Myoldmopar) [inline idd/Energy+.idd.in:108635]: Adds a few extra paths in the IDD, no problem.

File: scripts/dev/generate_epJSON_schema/modify_schema.py [hunk 4]
```diff
@@ -122,6 +122,7 @@ def isInt(s):
     'AirLoopHVAC:Splitter': 'nodes',
     'AirLoopHVAC:DedicatedOutdoorAirSystem': 'airloophvacs',
     'PythonPlugin:Variables': 'global_py_vars',
+    'PythonPlugin:SearchPaths': 'py_search_paths',
```
Comment (Myoldmopar) [inline scripts/dev/generate_epJSON_schema/modify_schema.py:125]: Names it `py_search_paths` for the input processor :heavy_check_mark:

File: src/EnergyPlus/PluginManager.cc [hunk 5]
```diff
@@ -62,6 +62,8 @@
 #include <EnergyPlus/PluginManager.hh>
 #include <EnergyPlus/UtilityRoutines.hh>
 
+#include <nlohmann/json.hpp>
```
Comment (Myoldmopar) [inline src/EnergyPlus/PluginManager.cc:65]: Includes the header to get access to the proper exception. :heavy_check_mark:

File: src/EnergyPlus/PluginManager.cc [hunk 6]
```diff
@@ -856,29 +858,38 @@ namespace PluginManagement {
                 auto const &fields = instance.value();
                 auto const &thisObjectName = instance.key();
                 inputProcessor->markObjectAsUsed(sPaths, thisObjectName);
-                std::string workingDirFlagUC = EnergyPlus::UtilityRoutines::MakeUPPERCase(fields.at("add_current_working_directory_to_search_path"));
+                std::string workingDirFlagUC = "YES";
+                try {
+                    workingDirFlagUC = EnergyPlus::UtilityRoutines::MakeUPPERCase(fields.at("add_current_working_directory_to_search_path"));
+                } catch (nlohmann::json::out_of_range &e) {
+                    // defaulted to YES
+                }
```
Comment (Myoldmopar) [inline src/EnergyPlus/PluginManager.cc:866]: Good fix here, if it was blank that would've been problematic.

File: src/EnergyPlus/PluginManager.cc [hunk 7]
```diff
@@ -856,29 +858,38 @@ namespace PluginManagement {
...
-                if (fields.find("search_path_3") != fields.end()) {
-                    PluginManager::addToPythonPath(PluginManager::sanitizedPath(fields.at("search_path_3")), true);
-                }
-                if (fields.find("search_path_4") != fields.end()) {
-                    PluginManager::addToPythonPath(PluginManager::sanitizedPath(fields.at("search_path_4")), true);
-                }
-                if (fields.find("search_path_5") != fields.end()) {
-                    PluginManager::addToPythonPath(PluginManager::sanitizedPath(fields.at("search_path_5")), true);
+                try {
+                    auto const vars = fields.at("py_search_paths");
```
Comment (Myoldmopar) [inline src/EnergyPlus/PluginManager.cc:881]: This lookup is OK since we are inside the try block.

File: src/EnergyPlus/PluginManager.cc [hunk 8]
```diff
@@ -856,29 +858,38 @@ namespace PluginManagement {
...
+                    for (const auto &var : vars) {
+                        try {
+                            std::cout << var.at("search_path");
```
Comment (Myoldmopar) [inline src/EnergyPlus/PluginManager.cc]: Probably need to remove this though.

File: src/EnergyPlus/PluginManager.cc [hunk 9]
```diff
@@ -1388,6 +1399,8 @@ namespace PluginManagement {
 #if LINK_WITH_PYTHON == 1
     void PluginManager::addToPythonPath(const std::string &path, bool userDefinedPath)
     {
+        if (path.empty()) return;
```
Comment (Myoldmopar) [inline src/EnergyPlus/PluginManager.cc:1401]: And here we don't try to add a blank path to the Python search path.  :heavy_check_mark:

**Review by Myoldmopar — CHANGES_REQUESTED** [review]
Stray debugging line was left in, otherwise this is good to go in.

**Comment by mitchute** [comment]
Sorry about that, should be good now. Why do something in 1 commit when you can spread it out over 10...

**Review by Myoldmopar — APPROVED** [review]

**Comment by Myoldmopar** [comment]
Pulled locally to confirm, everything builds and runs fine.  I'm going to merge this IDD change now.  Thanks @mitchute

**Pull request merged**